[GPU Process] [FormControls] Add a ControlPart for ProgressBar

shallawa opened the pull request:
#### 1674cd025bc6a390cb98dbe576137d3b4a24f873
<pre>
[GPU Process] [FormControls] Add a ControlPart for ProgressBar
<a href="https://bugs.webkit.org/show_bug.cgi?id=249675">https://bugs.webkit.org/show_bug.cgi?id=249675</a>
rdar://103569550

Reviewed by Aditya Keerthi.

The ProgressBarPart will handle drawing the progress bar form control. On macOS,
AppKit will be used to draw the platform control through the class ProgressBarMac.

ProgressBarPart will hold the attributes: &apos;position&apos; and &apos;animationStartTime&apos; which
control the display of the progress bar.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ControlPart.cpp:
(WebCore::ControlPart::rectForBounds):
* Source/WebCore/platform/graphics/controls/ControlPart.h:
* Source/WebCore/platform/graphics/controls/ControlStyle.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/controls/ControlStyle.h:
* Source/WebCore/platform/graphics/controls/PlatformControl.h:
(WebCore::PlatformControl::rectForBounds const):
* Source/WebCore/platform/graphics/controls/ProgressBarPart.cpp: Copied from Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h.(WebCore::ProgressBarPart::create):
(WebCore::ProgressBarPart::ProgressBarPart):
(WebCore::ProgressBarPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/ProgressBarPart.h: Copied from Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h.
(WebCore::ProgressBarPart::position const):
(WebCore::ProgressBarPart::animationStartTime const):
* Source/WebCore/platform/graphics/mac/controls/ButtonControlMac.mm:
(WebCore::ButtonControlMac::updateCellStates):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::createPlatformProgressBar):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::ControlMac::inflatedRect):
(WebCore::ControlMac::controlSizeForFont const):
(WebCore::ControlMac::controlSizeForSystemFont const):
(WebCore::ControlMac::controlSizeForSize const):
(WebCore::ControlMac::sizeForSystemFont const):
(WebCore::ControlMac::setFocusRingClipRect):
(WebCore::ControlMac::updateCellStates):
(WebCore::ControlMac::calculateControlSize const): Deleted.
* Source/WebCore/platform/graphics/mac/controls/MeterMac.h:
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.h: Copied from Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h.
(WebCore::ProgressBarMac::owningProgressBarPart const):
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm: Added.
(WebCore::ProgressBarMac::ProgressBarMac):
(WebCore::ProgressBarMac::cellSize const):
(WebCore::ProgressBarMac::cellOutsets const):
(WebCore::ProgressBarMac::rectForBounds const):
(WebCore::ProgressBarMac::draw):
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h:
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::createProgressBarPartForRenderer):
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::extractControlStyleStatesForRenderer const):
(WebCore::RenderTheme::progressBarRectForBounds const):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderThemeAdwaita.cpp:
(WebCore::RenderThemeAdwaita::progressBarRectForBounds const):
* Source/WebCore/rendering/RenderThemeAdwaita.h:
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
(WebCore::RenderThemeMac::progressBarRectForBounds const):
(-[WebCoreTextFieldCell _adjustedCoreUIDrawOptionsForDrawingBordersOnly:]): Deleted.
(-[WebCoreTextFieldCell _coreUIDrawOptionsWithFrame:inView:includeFocus:]): Deleted.
(-[WebCoreTextFieldCell _coreUIDrawOptionsWithFrame:inView:includeFocus:maskOnly:]): Deleted.
(WebCore::RenderThemeMac::progressBarSizes const): Deleted.
(WebCore::RenderThemeMac::progressBarMargins const): Deleted.
(WebCore::RenderThemeMac::minimumProgressBarHeight const): Deleted.
(WebCore::RenderThemeMac::paintProgressBar): Deleted.
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::encode):
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/258408@main">https://commits.webkit.org/258408@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a75f529ff23e243bbdad2dbdf742caf1fa84fba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101866 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11015 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34940 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111201 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105846 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11976 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1929 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/108958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107646 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/94274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/23853 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/4602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25339 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4687 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/1782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/10767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44826 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5763 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6439 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm:
```diff
@@ -0,0 +1,121 @@
+    auto& progressBarPart = owningProgressBarPart();
+    auto controlSize = calculateControlSize(style);
+    bool isIndeterminate = progressBarPart.position() < 0;
+    bool isActive = true; // isActive(renderObject)
```
This should read a bit from `ControlStyle`.

shallawa replied:
Fixed. We should be using
```
bool isActive = style.states.contains(ControlStyle::State::WindowActive);
```

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm:
```diff
@@ -0,0 +1,121 @@
+{
+}
+
+IntSize ProgressBarMac::cellSize(NSControlSize controlSize, const ControlStyle&) const
```
This method looks unused? Can we replace `RenderThemeMac::progressBarSizes`?

shallawa replied:
cellSize() is a virtual method which is called by `ControlMac::controlSizeForSize()`. Also `RenderThemeMac::progressBarSizes` is deleted.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm:
```diff
@@ -0,0 +1,121 @@
+    return sizes[controlSize];
+}
+
+IntOutsets ProgressBarMac::cellOutsets(NSControlSize controlSize, const ControlStyle&) const
```
Same comment as `cellSize`.

shallawa committed:
[GPU Process] [FormControls] Add a ControlPart for ProgressBar
https://bugs.webkit.org/show_bug.cgi?id=249675
rdar://103569550

Reviewed by Aditya Keerthi.

The ProgressBarPart will handle drawing the progress bar form control. On macOS,
AppKit will be used to draw the platform control through the class ProgressBarMac.

ProgressBarPart will hold the attributes: 'position' and 'animationStartTime' which
control the display of the progress bar.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ControlPart.cpp:
(WebCore::ControlPart::rectForBounds):
* Source/WebCore/platform/graphics/controls/ControlPart.h:
* Source/WebCore/platform/graphics/controls/ControlStyle.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/controls/ControlStyle.h:
* Source/WebCore/platform/graphics/controls/PlatformControl.h:
(WebCore::PlatformControl::rectForBounds const):
* Source/WebCore/platform/graphics/controls/ProgressBarPart.cpp: Copied from Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h.(WebCore::ProgressBarPart::create):
(WebCore::ProgressBarPart::ProgressBarPart):
(WebCore::ProgressBarPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/ProgressBarPart.h: Copied from Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h.
(WebCore::ProgressBarPart::position const):
(WebCore::ProgressBarPart::animationStartTime const):
* Source/WebCore/platform/graphics/mac/controls/ButtonControlMac.mm:
(WebCore::ButtonControlMac::updateCellStates):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::createPlatformProgressBar):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::ControlMac::inflatedRect):
(WebCore::ControlMac::controlSizeForFont const):
(WebCore::ControlMac::controlSizeForSystemFont const):
(WebCore::ControlMac::controlSizeForSize const):
(WebCore::ControlMac::sizeForSystemFont const):
(WebCore::ControlMac::setFocusRingClipRect):
(WebCore::ControlMac::updateCellStates):
(WebCore::ControlMac::calculateControlSize const): Deleted.
* Source/WebCore/platform/graphics/mac/controls/MeterMac.h:
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.h: Copied from Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h.
(WebCore::ProgressBarMac::owningProgressBarPart const):
* Source/WebCore/platform/graphics/mac/controls/ProgressBarMac.mm: Added.
(WebCore::ProgressBarMac::ProgressBarMac):
(WebCore::ProgressBarMac::cellSize const):
(WebCore::ProgressBarMac::cellOutsets const):
(WebCore::ProgressBarMac::rectForBounds const):
(WebCore::ProgressBarMac::draw):
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h:
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::createProgressBarPartForRenderer):
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::extractControlStyleStatesForRenderer const):
(WebCore::RenderTheme::progressBarRectForBounds const):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderThemeAdwaita.cpp:
(WebCore::RenderThemeAdwaita::progressBarRectForBounds const):
* Source/WebCore/rendering/RenderThemeAdwaita.h:
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
(WebCore::RenderThemeMac::progressBarRectForBounds const):
(-[WebCoreTextFieldCell _adjustedCoreUIDrawOptionsForDrawingBordersOnly:]): Deleted.
(-[WebCoreTextFieldCell _coreUIDrawOptionsWithFrame:inView:includeFocus:]): Deleted.
(-[WebCoreTextFieldCell _coreUIDrawOptionsWithFrame:inView:includeFocus:maskOnly:]): Deleted.
(WebCore::RenderThemeMac::progressBarSizes const): Deleted.
(WebCore::RenderThemeMac::progressBarMargins const): Deleted.
(WebCore::RenderThemeMac::minimumProgressBarHeight const): Deleted.
(WebCore::RenderThemeMac::paintProgressBar): Deleted.
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<ControlPart>::encode):
(IPC::ArgumentCoder<ControlPart>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/258408@main
